tmpauthenticator/__init__.py

```diff
@@ -30,19 +25,20 @@ async def get(self):
         login mechanism. This only happens when /tmplogin is hit - so you can use
         other parts of the hub as you normally would.
         """
-        # Create a new user.
+        username = str(uuid.uuid4())
+
+        # Run post_auth_hook
         #
-        # user_from_username ref: https://github.com/jupyterhub/jupyterhub/blob/4.0.0/jupyterhub/handlers/base.py#L504-L505
+        # Authenticator.run_post_auth_hook ref: https://github.com/jupyterhub/jupyterhub/blob/4.0.0/jupyterhub/auth.py#L400-L418
         #
-        username = str(uuid.uuid4())
-        user = self.user_from_username(username)
+        authenticated = {"name": username}
+        authenticated = await self.authenticator.run_post_auth_hook(self, authenticated)
 
-        # Let a subclasses of TmpAuthenticator process the new user by
-        # overriding TmpAuthenticator.process_user.
+        # Create a new user
+        #
+        # BaseHandler.auth_to_user ref: https://github.com/jupyterhub/jupyterhub/blob/4.0.0/jupyterhub/handlers/base.py#L774-L821
         #
-        user = self.process_user(user, self)
-        if inspect.isawaitable(user):
-            user = await user
+        user = await self.auth_to_user(authenticated)
```

Comment:
It probably makes the most to call the outer wrapper `authenticator.get_authenticated_user`. This is the method that applies block lists, validation, runs hooks, etc.

Then it would be the responsibility of `authenticate` to generate the new name `if isinstance(handler, TmpAuthenticateHandler)`

Reply:
Hmmm but `get_authenticated_user` doesn't provide a user, only an auth dictionary. We need the user object to do `set_hub_cookie`, which must be called explicitly to get the intented behavior of getting a new user setup whenever /hub/tmplogin is visited.

I tried and got things working by using `BaseHandler.login_user` if you think that is cleaner. Here is the logic overview for that:

- [`BaseHandler.login_user`](https://github.com/jupyterhub/jupyterhub/blob/4.0.0/jupyterhub/handlers/base.py#L823-L843)
  - calls BaseHandler.authenticate which just calls [`Authenticator.get_authenticated_user`](https://github.com/jupyterhub/jupyterhub/blob/4.0.0/jupyterhub/auth.py#L472-L534)
  - calls auth_to_user
  - calls set_login_cookie
    - conditionally calls set_session_cookie
    - conditionally calls set_hub_cookie (we want to do this unconditionally)

---

The logic in the /hub/tmplogin handler then becomes

```python
user = await self.login_user(None)
self.set_hub_cookie(user)
```

And we need to add TmpAuthenticator.authenticate like this as well:

```python
    async def authenticate(self, handler, data):
        """FIXME: write docstring"""
        username = str(uuid.uuid4())
        return {
            "name": username,
        }
```

Reply:
> It's unclear to me what tmplogin does that standard login customization can't (login is more customizable now than when tmpauthenticator started).

I'm not sure what technicques you think of as standard login customization, but a key thing that may be tricky is to ignore an existing user as recognized by a cookie and get a new one, and to allow for users to write URLs like `/hub/tmplogin?next=...` that they know they can visit and get a new user.

Reply:
Ah, right - that behavior of always creating a new user whether you are logged in or not would probably be hard to replicate without at least a LoginHandler subclass.